app/src/pages/playground/Playground.tsx

```diff
@@ -208,17 +208,24 @@ const DEFAULT_EXPANDED_PROMPTS = ["prompts"];
 const DEFAULT_EXPANDED_PARAMS = ["input", "output"];
 
 function PlaygroundContent() {
-  const instances = usePlaygroundContext((state) => state.instances);
   const templateFormat = usePlaygroundContext((state) => state.templateFormat);
   const [searchParams] = useSearchParams();
   const datasetId = searchParams.get("datasetId");
   const isDatasetMode = datasetId != null;
-  const numInstances = instances.length;
+  const numInstances = usePlaygroundContext((state) => state.instances.length);
   const isSingleInstance = numInstances === 1;
-  const isRunning = instances.some((instance) => instance.activeRunId != null);
-  const anyDirtyPromptInstances = instances
-    .filter((i) => i.prompt)
-    .some((instance) => instance.dirty);
+  const isRunning = usePlaygroundContext((state) =>
+    state.instances.some((instance) => instance.activeRunId != null)
+  );
+  const anyDirtyPromptInstances = usePlaygroundContext((state) =>
+    Object.values(state.dirtyInstances).some((dirty) => dirty)
+  );
+  const instanceIds = usePlaygroundContext(
+    (state) => state.instances.map((instance) => instance.id),
+    (left, right) =>
+      left.length === right.length &&
+      left.every((id, index) => id === right[index])
```

Comment:
what's this block do?

Reply:
custom equality fn for this selector so that it only triggers a re-render when the actual contents of this array changes

Reply:
Added comment